examples/smc.jl

```diff
@@ -4,32 +4,45 @@ using Distributions
 using Plots
 using StatsFuns
 
-# Particle Filter implementation
-struct Particle{T} # Here we just need a tree
+# Particle Filter 
+struct Particle{T<:AbstractStateSpaceModel,V} <: AbstractParticle{T}
     parent::Union{Particle,Nothing}
-    state::T
+    state::V
 end
 
-Particle(state::T) where {T} = Particle(nothing, state)
-Particle() = Particle(nothing, nothing)
-Base.show(io::IO, p::Particle) = print(io, "Particle($(p.state))")
+function Particle(parent, state, model::T) where {T<:AbstractStateSpaceModel}
+    return Particle{T,particleof(model)}(parent, state)
+end
+
+function Particle(model::T) where {T<:AbstractStateSpaceModel}
+    N = dimension(model)
+    V = particleof(model)
+    state = N == 1 ? zero(V) : zeros(V, N)
+    return Particle{T,V}(nothing, state)
+end
+
+Base.show(io::IO, p::Particle{T,V}) where {T,V} = print(io, "Particle{$T, $V}($(p.state))")
+
+const ParticleContainer{T} = AbstractVector{<:Particle{T}}
 
 """
     linearize(particle)
 
 Return the trace of a particle, i.e. the sequence of states from the root to the particle.
 """
-function linearize(particle::Particle{T}) where {T}
-    trace = T[]
-    parent = particle.parent
-    while !isnothing(parent)
-        push!(trace, parent.state)
-        parent = parent.parent
+function linearize(particle::Particle{T,V}) where {T,V}
+    trace = V[]
+    current = particle
+    while !isnothing(current)
+        push!(trace, current.state)
+        current = current.parent
     end
-    return trace
+    return trace[1:(end - 1)] # Discard the root node
 end
 
-ParticleContainer = AbstractVector{<:Particle}
+function isdone(t, model::AbstractStateSpaceModel, particles::ParticleContainer)
+    return all(map(particle -> isdone(t, model, particle), particles))
```

Comment:
Without intermediate allocations1:
```suggestion
    return all(particle -> isdone(t, model, particle), particles)
```